### ISSUE TITLE
[Bug] Multi-GPU hangs on medium to big datasets and with MMClsWandbHook

### ISSUE THREAD

**Issue by robin-maillot** [open-mmlab/mmclassification#1233]
### Branch

master branch (0.24 or other 0.x version)

### Describe the bug

When running pytorch in distributed mode over 2 gpus training hangs after one epoch before reporting the validation loss. It then timeouts after 30min. The same code runs fine on a single GPU and the validation step takes <<30min

## Logs (sorry about the image)

![image](https://user-images.githubusercontent.com/26007590/204510299-a2f48d98-bbd6-4a1b-9274-da0dd8f5a71c.png)

## Example Config file (there is some custom stuff for building the dataset but normally nothing special that runs during validation):
```
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True
)
train_pipeline = [
    dict(type="LoadImageFromFile"),
    dict(type="RandomFlip", flip_prob=0.5, direction="horizontal"),
    dict(type="RandomFlip", flip_prob=0.5, direction="vertical"),
    dict(type="Albu", transforms=[dict(type="Rotate", limit=90, p=0.5)]),
    dict(
        type="Normalize",
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True,
    ),
    dict(type="ImageToTensor", keys=["img"]),
    dict(type="ToTensor", keys=["gt_label"]),
    dict(type="Collect", keys=["img", "gt_label"]),
]
test_pipeline = [
    dict(type="LoadImageFromFile"),
    dict(
        type="Normalize",
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True,
    ),
    dict(type="ImageToTensor", keys=["img"]),
    dict(type="Collect", keys=["img"]),
]
data = dict(
    samples_per_gpu=16,
    workers_per_gpu=2,
    train=dict(
        type="ClassBalancedDataset",
        oversample_thr=10,
        dataset=dict(
            type=dataset_type,
            dataset_type="train",
            dataset_specs=DATASETS_SPECS,
            filter=FILTER,
            canvaser=CANVASER,
            pipeline=train_pipeline,
        ),
        pipeline=train_pipeline,
    ),
    val=dict(
        type=dataset_type,
        dataset_type="val",
        dataset_specs=DATASETS_SPECS,
        filter=FILTER,
        canvaser=CANVASER,
        pipeline=test_pipeline,
    ),
    test=dict(
        type=dataset_type,
        dataset_type="test",
        dataset_specs=DATASETS_SPECS,
        filter=FILTER,
        canvaser=CANVASER,
        pipeline=test_pipeline,
    ),
)
evaluation = dict(interval=1, metric="f1_score")
log_level = "INFO"
checkpoint_config = dict(interval=1)
dist_params = dict(backend="nccl")
resume_from = None
lr_config = dict(policy="fixed")
workflow = [("train", 1), ("val", 1)]
load_from = "https://download.openmmlab.com/mmclassification/v0/resnet/resnet34_batch256_imagenet_20200708-32ffb4f7.pth"
model = dict(
    type="ImageClassifier",
    backbone=dict(
        type="ResNet", depth=34, num_stages=4, out_indices=(3,), style="pytorch"
    ),
    neck=dict(type="GlobalAveragePooling"),
    head=dict(
        type="LinearClsHead",
        in_channels=512,
        loss=dict(type="CrossEntropyLoss", loss_weight=1.0),
    ),
)
custom_hooks = [
    dict(
        type="EarlyStoppingHook",
        monitor="loss",
        phase="val",
        patience=5,
        verbose=True,
        mode="min",
    )
]
optimizer = dict(type="Adam", lr=0.01, eps=0.1)
optimizer_config = dict(grad_clip=dict(max_norm=1.0, norm_type=2))
runner = dict(type="EpochBasedRunner", max_epochs=200)
```

The arguments specified when running on 2 GPUs was:
```
NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=GRAPH python3 -m torch.distributed.launch --nnodes=1 --node_rank=0 --master_addr="127.0.0.1" --nproc_per_node=2 --master_port=29500 <custom_train_path.py> <custom_config_path.py> --launcher pytorch
```

The issue happened when running on a dataset > 10000 images and when using a small dataset (1000 images) the issue disapeared

### Environment

```
{'sys.platform': 'linux',
 'Python': '3.8.10 (default, Mar 15 2022, 12:22:08) [GCC 9.4.0]',
 'CUDA available': True,
 'GPU 0,1': 'Tesla T4',
 'CUDA_HOME': '/usr/local/cuda',
 'NVCC': 'Cuda compilation tools, release 11.3, V11.3.109',
 'GCC': 'x86_64-linux-gnu-gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0',
 'PyTorch': '1.11.0+cu113',
 'PyTorch compiling details': 'PyTorch built with:\n'
                              '  - GCC 7.3\n'
                              '  - C++ Version: 201402\n'
                              '  - Intel(R) Math Kernel Library Version '
                              '2020.0.0 Product Build 20191122 for Intel(R) 64 '
                              'architecture applications\n'
                              '  - Intel(R) MKL-DNN v2.5.2 (Git Hash '
                              'a9302535553c73243c632ad3c4c80beec3d19a1e)\n'
                              '  - OpenMP 201511 (a.k.a. OpenMP 4.5)\n'
                              '  - LAPACK is enabled (usually provided by '
                              'MKL)\n'
                              '  - NNPACK is enabled\n'
                              '  - CPU capability usage: AVX2\n'
                              '  - CUDA Runtime 11.3\n'
                              '  - NVCC architecture flags: '
                              '-gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86\n'
                              '  - CuDNN 8.2\n'
                              '  - Magma 2.5.2\n'
                              '  - Build settings: BLAS_INFO=mkl, '
                              'BUILD_TYPE=Release, CUDA_VERSION=11.3, '
                              'CUDNN_VERSION=8.2.0, '
                              'CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, '
                              'CXX_FLAGS= -Wno-deprecated '
                              '-fvisibility-inlines-hidden -DUSE_PTHREADPOOL '
                              '-fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM '
                              '-DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK '
                              '-DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE '
                              '-DEDGE_PROFILER_USE_KINETO -O2 -fPIC '
                              '-Wno-narrowing -Wall -Wextra '
                              '-Werror=return-type '
                              '-Wno-missing-field-initializers '
                              '-Wno-type-limits -Wno-array-bounds '
                              '-Wno-unknown-pragmas -Wno-sign-compare '
                              '-Wno-unused-parameter -Wno-unused-function '
                              '-Wno-unused-result -Wno-unused-local-typedefs '
                              '-Wno-strict-overflow -Wno-strict-aliasing '
                              '-Wno-error=deprecated-declarations '
                              '-Wno-stringop-overflow -Wno-psabi '
                              '-Wno-error=pedantic -Wno-error=redundant-decls '
                              '-Wno-error=old-style-cast '
                              '-fdiagnostics-color=always -faligned-new '
                              '-Wno-unused-but-set-variable '
                              '-Wno-maybe-uninitialized -fno-math-errno '
                              '-fno-trapping-math -Werror=format '
                              '-Wno-stringop-overflow, LAPACK_INFO=mkl, '
                              'PERF_WITH_AVX=1, PERF_WITH_AVX2=1, '
                              'PERF_WITH_AVX512=1, TORCH_VERSION=1.11.0, '
                              'USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, '
                              'USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, '
                              'USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, '
                              'USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, \n',
 'TorchVision': '0.12.0+cu113',
 'OpenCV': '4.6.0',
 'MMCV': '1.6.0',
 'MMCV Compiler': 'GCC 9.3',
 'MMCV CUDA Compiler': '11.3',
 'MMClassification': '0.23.2+a6c2aec'}
```

### Other information

Some people seem to have had similar issues with but none of these fixes helped:
- [pytorch issue](https://github.com/pytorch/pytorch/issues/77562)
- [huggingface issue](https://github.com/huggingface/accelerate/issues/314)
- 
I find it strange that no one has had issues with mmcls and multi-gpu. This makes me think this is probably a bug with my code and not the actual mmcls code.

**Comment by Ezra-Yu** [comment]
Is this always the case or is it random?  

Have you tried only setting `workflow = [("train", 1)]`?

**Comment by robin-maillot** [comment]
Yep just tried that and I get the exact same failure (with the val f-score being reported first)

**Comment by robin-maillot** [comment]
@Ezra-Yu I managed to recreate this using the following config file (after downloading the cats and dogs dataset). Removing wandb logger fixed the hanging for me. I have had issues with wandb and multi gpu before, do you think that is where the issue can come from?

I am trying without the wandb logger with our custom data in the meantime to see if it solves the issue

# Info
## Download dataset

```
# Download the cats & dogs dataset
cd ~/.
!wget https://www.dropbox.com/s/wml49yrtdo53mie/cats_dogs_dataset_reorg.zip?dl=0 -O cats_dogs_dataset.zip
!mkdir -p data
!unzip -qo cats_dogs_dataset.zip -d ./data/
```

## Config File

```
mmcls_root = "/.venv/lib/python3.8/site-packages/mmcls/.mim"
_base_ = [
    f"{mmcls_root}/configs/_base_/models/resnet50.py",
    f"{mmcls_root}/configs/_base_/default_runtime.py",
]

# model settings
model = dict(
    backbone=dict(
        init_cfg=dict(
            type="Pretrained",
            checkpoint="https://download.openmmlab.com/mmclassification/v0/resnet/resnet50_8xb32_in1k_20210831-ea4938fc.pth",
            prefix="backbone",
        )
    ),
    head=dict(num_classes=2),
)
# evaluation = dict(interval=1, metric="accuracy", metric_options={"topk": (1,)})
evaluation = dict(interval=1, metric="f1_score")

# optimizer
optimizer = dict(type="SGD", lr=0.01, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=None)
# learning policy
lr_config = dict(policy="step", step=[15])
# checkpoint saving
checkpoint_config = dict(interval=1)
# yapf:disable
log_config = dict(
    interval=100,
    hooks=[
        dict(type="TextLoggerHook"),
        dict(
            type="MMClsWandbHook",
            init_kwargs={
                "project": "mmcls",
                "name": "will-be-added-by-train-script",
            },
            interval=100,
            log_checkpoint=False,
            log_checkpoint_metadata=False,
            num_eval_images=100,
        ),
    ],
)
# yapf:enable
# runtime settings
runner = dict(type="EpochBasedRunner", max_epochs=50)
dist_params = dict(backend="nccl")
log_level = "INFO"
work_dir = "./work_dirs/mnist/"
load_from = None
resume_from = None
workflow = [("train", 1), ("val", 1)]


# dataset settings
dataset_type = "ImageNet"  # dataset name，
img_norm_cfg = dict(  # Image normalization config to normalize the input images
    mean=[
        123.675,
        116.28,
        103.53,
    ],  # Mean values used to pre-training the pre-trained backbone models
    std=[
        58.395,
        57.12,
        57.375,
    ],  # Standard variance used to pre-training the pre-trained backbone models
    to_rgb=True,
)  # Whether to invert the color channel, rgb2bgr or bgr2rgb.
# train data pipeline
train_pipeline = [
    dict(type="LoadImageFromFile"),  # First pipeline to load images from file path
    dict(type="RandomResizedCrop", size=224),  # RandomResizedCrop
    dict(
        type="RandomFlip", flip_prob=0.5, direction="horizontal"
    ),  # Randomly flip the picture horizontally with a probability of 0.5
    dict(type="Normalize", **img_norm_cfg),  # normalization
    dict(
        type="ImageToTensor", keys=["img"]
    ),  # convert image from numpy into torch.Tensor
    dict(type="ToTensor", keys=["gt_label"]),  # convert gt_label into torch.Tensor
    dict(
        type="Collect", keys=["img", "gt_label"]
    ),  # Pipeline that decides which keys in the data should be passed to the detector
]
# test data pipeline
test_pipeline = [
    dict(type="LoadImageFromFile"),
    dict(type="Resize", size=(256, -1)),
    dict(type="CenterCrop", crop_size=224),
    dict(type="Normalize", **img_norm_cfg),
    dict(type="ImageToTensor", keys=["img"]),
    dict(type="Collect", keys=["img"]),  # do not pass gt_label while testing
]
data = dict(
    samples_per_gpu=32,  # Batch size of a single GPU
    workers_per_gpu=2,  # Worker to pre-fetch data for each single GPU
    train=dict(  # train data config
        type=dataset_type,  # dataset name
        data_prefix="~/data/cats_dogs_dataset/training_set/training_set",  # Dataset root, when ann_file does not exist, the category information is automatically obtained from the root folder
        classes="~/data/cats_dogs_dataset/classes.txt",
        pipeline=train_pipeline,
    ),  # train data pipeline
    val=dict(  # val data config
        type=dataset_type,
        data_prefix="~/data/cats_dogs_dataset/val_set/val_set",  # Dataset root, when ann_file does not exist, the category information is automatically obtained from the root folder
        ann_file="~/data/cats_dogs_dataset/val.txt",  #  ann_file existes, the category information is obtained from file
        classes="~/data/cats_dogs_dataset/classes.txt",
        pipeline=test_pipeline,
    ),
    test=dict(  # test data config
        type=dataset_type,
        data_prefix="~/data/cats_dogs_dataset/test_set/test_set",  # Dataset root, when ann_file does not exist, the category information is automatically obtained from the root folder
        ann_file="~/data/cats_dogs_dataset/test.txt",  #  ann_file existes, the category information is obtained from file
        classes="~/data/cats_dogs_dataset/classes.txt",
        pipeline=test_pipeline,
    ),
)
```

## Command used:
```
./tools/dist_train.sh <CONFIG_FILE_PATH> 2
```

**Comment by ssumoo** [comment]
@Ezra-Yu I can help but I'll need guidance to get started. 

How do you guys normally debug multi-gpu issues? I'm stuck with this issue as well.

**Comment by Ezra-Yu** [comment]
@ssumoo  We have also been having headaches with multi-GPU bugs, generally using logs as well as hooks.

For this issue,  Here is probably because Wandbhook,  first, confirms whether it is in the train or eval phase, to see whether there is a saved checkpoint. Then, insert some 'print' or 'log' in F such as  ``after_run(self, runner):```, ```after_train_epoch``` , into and return to the function, to see exactly which section of code the problem is.

**Comment by mzr1996** [comment]
Get it, it's caused by a wrong `@master_only`, and I have fixed and tested it in #1242
